tools/kv/keyvaluestore.go

```diff
@@ -0,0 +1,204 @@
+func (k *Store) Save() error {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	// Short-circuit if we have no changes
+	if len(k.log) == 0 {
+		return nil
+	}
+
+	// firstly peek at version; open read/write to keep lock between check & write
+	f, err := os.OpenFile(k.filename, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return err
+	}
+	stat, _ := os.Stat(k.filename)
+
+	defer f.Close()
+
+	// Only try to merge if > 0 bytes, ignore empty files (decoder will fail)
+	if stat.Size() > 0 {
+		k.loadAndMergeReaderIfNeeded(f)
+		// Now we overwrite the file
+		f.Seek(0, os.SEEK_SET)
+		f.Truncate(0)
+	}
+
+	k.version++
+
+	enc := gob.NewEncoder(f)
+	if err := enc.Encode(k.version); err != nil {
+		return fmt.Errorf("Error while writing version data to %v: %v", k.filename, err)
+	}
+	if err := enc.Encode(k.db); err != nil {
+		return fmt.Errorf("Error while writing new key/value data to %v: %v", k.filename, err)
+	}
+	// Clear log now that it's saved
+	k.log = nil
+
+	return nil
+}
+
+// Reads as little as possible from the passed in file to determine if the
+// contents are different from the version already held. If so, reads the
+// contents and merges with any outstanding changes. If not, stops early without
+// reading the rest of the file
+func (k *Store) loadAndMergeIfNeeded() error {
+	stat, err := os.Stat(k.filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil // missing is OK
+		}
+		return err
+	}
+	// Do nothing if empty file
+	if stat.Size() == 0 {
+		return nil
+	}
+
+	f, err := os.OpenFile(k.filename, os.O_RDONLY, 0664)
+	if err == nil {
+		defer f.Close()
+		return k.loadAndMergeReaderIfNeeded(f)
+	} else {
+		return err
+	}
+}
+
+// As loadAndMergeIfNeeded but lets caller decide how to manage file handles
+func (k *Store) loadAndMergeReaderIfNeeded(f io.Reader) error {
+	var versionOnDisk int64
+	// Decode *only* the version field to check whether anyone else has
+	// modified the db; gob serializes structs in order so it will always be 1st
+	dec := gob.NewDecoder(f)
+	err := dec.Decode(&versionOnDisk)
+	if err != nil {
+		return fmt.Errorf("Problem checking version of key/value data from %v: %v", k.filename, err)
+	}
+	// Totally uninitialised Version == 0, saved versions are always >=1
+	if versionOnDisk != k.version {
+		// Reload data & merge
+		var dbOnDisk map[string]interface{}
+		err = dec.Decode(&dbOnDisk)
+		if err != nil {
+			return fmt.Errorf("Problem reading updated key/value data from %v: %v", k.filename, err)
+		}
+		k.reapplyChanges(dbOnDisk)
+		k.version = versionOnDisk
+	}
+	return nil
+}
+
+// reapplyChanges replays the changes made since the last load onto baseDb
+// and stores the result as our own DB
+func (k *Store) reapplyChanges(baseDb map[string]interface{}) {
+	for _, change := range k.log {
+		switch change.op {
+		case setOperation:
+			baseDb[change.key] = change.value
+		case removeOperation:
+			delete(baseDb, change.key)
```

Comment:
There's a bit of duplication here with `Set()` and `Remove()`. Though since this is just a wrapper around a map, it's probably fine for now. One way to solve it is to turn `change` into an interface, and calling `Apply()` on `*setKey` or `*removeKey` implementations in `reapplyChanges()` and `logChange()`.

It looks like a mutex is needed here, but I think it's fine this code path is only called on `Save()` (which has a mutex) and `NewStore()` (which shouldn't need one).